usr/src/uts/common/fs/zfs/zap.c

```diff
@@ -505,6 +505,16 @@ zap_get_leaf_byblk(zap_t *zap, uint64_t blkid, dmu_tx_t *tx, krw_t lt,
 
 	ASSERT(RW_LOCK_HELD(&zap->zap_rwlock));
 
+	/*
+	 * If system crashed just after dmu_free_long_range in zfs_rmnode, we
+	 * would be left with an empty xattr dir in delete queue. blkid=0
+	 * would be passed in when doing zfs_purgedir. If that's the case we
+	 * should just return immediately. The underlying objects should
+	 * already be freed, so this should be perfectly fine.
+	 */
+	if (blkid == 0)
```

Comment:
zap_lockdir() definitely happens before you can do anything else with the ZAP object.  It looks like mzap_open(), near where it checks for ZBT_MICRO, is the right place to add the checks I mentioned above.

I'd be fine with adding something like the checks you mentioned in zap_deref_leaf(), to check that the leaf also has ZAP_MAGIC and ZBT_LEAF.  But mzap_open() should be our first line of defense.